### PR TITLE
[DON'T MERGE]Task/generic aggregation

### PR DESCRIPTION
**Pull request by IvanHdzC**
This PR is an optimization for closed issue #1787. The reason for this optimization is because with this PR more sinks share the aggregation process.

The Generic aggregation class is now used as a bean and it declares (it's an abstract class) two functions for child classes.

The Generic Column and Row aggregation classes implements initialize and aggregation methods.

All data processing function are declared on Utils class in order to be shared among all sinks. If any function is very specific for a sink, it's declared on the sink itself.

Functional testing is still pending, but it would be nice if you guys can take a look to the general idea and implementation.

### CODE REVIEW

File: cygnus-ngsi/src/main/java/com/telefonica/iot/cygnus/aggregation/NGSIGenericAggregator.java [hunk 1]
```diff
@@ -73,225 +49,282 @@
...
+
+    public boolean isAttrNativeTypes() {
+        return attrNativeTypes;
+    }
+
+    public void setAttrNativeTypes(boolean attrNativeTypes) {
+        this.attrNativeTypes = attrNativeTypes;
+    }
+
+    public boolean isEnableGrouping() {
```
Comment (fgalan) [inline cygnus-ngsi/src/main/java/com/telefonica/iot/cygnus/aggregation/NGSIGenericAggregator.java]: Maybe it is to be done in a next commit, but the style has somehow broken... Functions should have a JavaDoc description and the blocks an ending `// comment`.

File: cygnus-ngsi/src/test/java/com/telefonica/iot/cygnus/sinks/NGSIMongoSinkTest.java [hunk 2]
```diff
@@ -0,0 +1,259 @@
+package com.telefonica.iot.cygnus.sinks;
```
Comment (fgalan) [inline cygnus-ngsi/src/test/java/com/telefonica/iot/cygnus/sinks/NGSIMongoSinkTest.java]: License header in new files

**Comment by fgalan** [comment]
This PR involves a large refactor so I wonder how it may impact to other ongoing PRs. In particular, to PR https://github.com/telefonicaid/fiware-cygnus/pull/1817. Could you analyze?

File: cygnus-ngsi/src/test/java/com/telefonica/iot/cygnus/sinks/NGSIMongoSinkTest.java (hunk 2 above)
Comment (IvanHdzC) [reply cygnus-ngsi/src/test/java/com/telefonica/iot/cygnus/sinks/NGSIMongoSinkTest.java]: Fixed in [25f98f6](https://github.com/telefonicaid/fiware-cygnus/pull/1818/commits/25f98f637db431e38bfee19706ae5ac09993eb24)

File: cygnus-ngsi/src/main/java/com/telefonica/iot/cygnus/aggregation/NGSIGenericAggregator.java (hunk 1 above)
Comment (IvanHdzC) [reply cygnus-ngsi/src/main/java/com/telefonica/iot/cygnus/aggregation/NGSIGenericAggregator.java]: Fixed in [683ab07](https://github.com/telefonicaid/fiware-cygnus/commit/683ab076f020f01987369925c4e9091d083b7560)

File: cygnus-ngsi/src/main/java/com/telefonica/iot/cygnus/sinks/NGSIHDFSSink.java [hunk 3]
```diff
@@ -18,6 +18,12 @@
 
 package com.telefonica.iot.cygnus.sinks;
 
+import com.amazonaws.services.dynamodbv2.xspec.S;
```
Comment (AlvaroVega) [inline cygnus-ngsi/src/main/java/com/telefonica/iot/cygnus/sinks/NGSIHDFSSink.java]: Used by?

File: cygnus-ngsi/src/test/java/com/telefonica/iot/cygnus/sinks/NGSIHDFSSinkTest.java [hunk 4]
```diff
@@ -1533,76 +1615,101 @@ public void testNativeTypeRowBatchCsvNotMetadata() throws CygnusBadConfiguration
...
-        NGSIHDFSSink ngsihdfsSink= new NGSIHDFSSink();
-        ngsihdfsSink.configure(createContextforNativeTypes("json-row", null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, null, attr_native_types, "true"));
         NGSIBatch batch = prepaireBatch();
         String destination = "someDestination";
+        String file_format = "json-row";
+        NGSIHDFSSink ngsihdfsSink = new NGSIHDFSSink();
+        ngsihdfsSink.configure(createContextforNativeTypes(null, null, null, null, null, null, null, null, null, null, null, file_format, null, null, null, null, null, null, null, null, null, null));
         try {
             batch.startIterator();
+            //NGSIGenericAggregator aggregator = new NGSIGenericColumnAggregator();
```
Comment (AlvaroVega) [inline cygnus-ngsi/src/test/java/com/telefonica/iot/cygnus/sinks/NGSIHDFSSinkTest.java]: leftover?

**Comment by IvanHdzC** [comment]
> This PR involves a large refactor so I wonder how it may impact to other ongoing PRs. In particular, to PR #1817. Could you analyze?

For this case, it wouldn't be necessary to re work. Since on the last commit of that PR the method `buildPkgName` was modifyed to use the name of the entity or name it the usuall way. Those methods are used as well on the new version. Maybe it would be necessary to some little changes but it shouldn't be that complicated.

File: cygnus-ngsi/src/main/java/com/telefonica/iot/cygnus/sinks/NGSIHDFSSink.java (hunk 3 above)
Comment (IvanHdzC) [reply cygnus-ngsi/src/main/java/com/telefonica/iot/cygnus/sinks/NGSIHDFSSink.java]: No, there isn't any new dependency, not very sure why I added it, but it's never used.

Fixed in [bf42de0](https://github.com/telefonicaid/fiware-cygnus/pull/1818/commits/bf42de008c59fb41d740ec6b851aee434e50fb7b)

File: cygnus-ngsi/src/test/java/com/telefonica/iot/cygnus/sinks/NGSIHDFSSinkTest.java (hunk 4 above)
Comment (IvanHdzC) [reply cygnus-ngsi/src/test/java/com/telefonica/iot/cygnus/sinks/NGSIHDFSSinkTest.java]: Fixed in [bf42de0](https://github.com/telefonicaid/fiware-cygnus/pull/1818/commits/bf42de008c59fb41d740ec6b851aee434e50fb7b)

**Comment by IvanHdzC** [comment]
### functional testing in progress

**Comment by mrutid** [comment]
Parece que hemos introducido un bug. Cuando un atributo tiene un nombre que contiene "_type" parece que no se persiste convenientemente al menos el HDFS. Posiblemente pase en más sinks y esta solución no sea válida. Lo comprobamos.

Usar el magicString "_type" debneria habernos mosqueado un poco.

**Pull request merged**